# Group 6 SC1015 Assignment

We will be looking into Car Features and how it will affect the MSRP of the cars. <br>

This dataset was from kaggle [Car Features and MSRP](https://www.kaggle.com/CooperUnion/cardataset) <br>

Since a description of the data was not given. Let us list out the features and their details <br>
  - __Car Make__
      * Brand of the vehicle
  - __Car Model__
      * Can also be "Name" of car
  - __Car Year__
      * When the car was manufactured
  - __Engine Fuel Type__
      * Fuel type reccomendation for the car. If it is an electric vehicle "electric" will be shown.
  - __Engine Hourse Power(HP)__
      * Refers to the power an engine produces
  - __Engine Cylinders__
      * Cylinders are located in an engine, it is  a chamber where fuel is combusted and power is generated. In this data, it refers to the amount of cylinders the Engine has.
  - __Transimission Type__
      * The way power is transmitted from an engine to the axle in the vehicle
  - __Driven Wheels__
      * Different Vehicles run on different wheel drives. Front wheel drive cars will have power transmitted to the Front 2 wheels. Rear wheel drive cars will have power transmitted to the Rear 2 wheels. All wheel drive cars will have power transmitted to all the Wheels all the time. 4WD is similar to All wheel drive cars, just with the fact that it can toggle between 4WD and 2WD. 
  - __Number of Doors__
      * Self explanatory
  - __Market Category__
      * Refers to the market audience for the cars
  - __Vehicle Size__
      * Self explanatory
  - __Vehicle Style__
      * Refers to their body style, We are familiar with cars without roofs (Convertible), normal cars (Sedans) and many more
  - __Highway MPG__
      * Refers to Miles Per Gallons, and it is used to show how far the car can travel for every gallon on an highway. (Travelling on highway does not require the car to Stop and Accelerate constantly. Therefore will usually result in Higher MPG compared to City MPG)
  - __City MPG__
      * Similar to Highway MPG, but shows how far the car travels for every gallon in the City.
  - __Popularity__
      * Data was scraped from twitter polls to indicate the popularity of the car.
  - __MSRP__
      * Refers to the manufacturer's suggested retail price
    

Find Link to Sections here <br>
   - [1. Import Libraries](#Importing)
   - [2. Data Extraction](#Data)
   - [3. Checking Data](#Check_Data)
   - [4. Performing EDA](#EDA)
  

<hr>

# 1. Import Libraries

> Importing libraries for EDA

<a id = "Importing"></a>

In [99]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

<hr>

# 2. Data Extraction 

<a id = "Data"></a>

In [100]:
car_raw = pd.read_csv('Car.csv')

car_raw.sample(10)

Make           Model  Year                Engine Fuel Type  \
9540   Chevrolet  Silverado 1500  2015        flex-fuel (unleaded/E85)   
1983   Chevrolet         Bolt EV  2017                        electric   
3466       Dodge          Dakota  2009                regular unleaded   
1221       Honda          Accord  2015                regular unleaded   
966      Porsche             911  2017     premium unleaded (required)   
5671  Volkswagen             GTI  2013  premium unleaded (recommended)   
1497    Cadillac       ATS Coupe  2015        flex-fuel (unleaded/E85)   
5268    Maserati          Ghibli  2015     premium unleaded (required)   
9727     Hyundai          Sonata  2015                regular unleaded   
6226      Nissan            Juke  2015  premium unleaded (recommended)   

      Engine HP  Engine Cylinders Transmission Type      Driven_Wheels  \
9540      285.0               6.0         AUTOMATIC   rear wheel drive   
1983      200.0               NaN      DIRECT_DRIVE  front wheel drive   
3466      210.0               6.0         AUTOMATIC   rear wheel drive   
1221      185.0               4.0            MANUAL  front wheel drive   
966       580.0               6.0  AUTOMATED_MANUAL    all wheel drive   
5671      200.0               4.0            MANUAL  front wheel drive   
1497      321.0               6.0         AUTOMATIC    all wheel drive   
5268      345.0               6.0         AUTOMATIC   rear wheel drive   
9727      245.0               4.0         AUTOMATIC  front wheel drive   
6226      188.0               4.0            MANUAL  front wheel drive   

      Number of Doors                    Market Category Vehicle Size  \
9540              2.0                          Flex Fuel        Large   
1983              4.0                          Hatchback      Compact   
3466              4.0                          Flex Fuel      Compact   
1221              4.0                                NaN      Midsize   
966               2.0            Luxury,High-Performance      Compact   
5671              4.0              Hatchback,Performance      Compact   
1497              2.0  Flex Fuel,Luxury,High-Performance      Compact   
5268              4.0     Exotic,Luxury,High-Performance      Midsize   
9727              4.0                        Performance      Midsize   
6226              4.0    Crossover,Hatchback,Performance      Compact   

            Vehicle Style  highway MPG  city mpg  Popularity    MSRP  
9540   Regular Cab Pickup           24        18        1385   33215  
1983        4dr Hatchback          110       128        1385   40905  
3466  Extended Cab Pickup           20        15        1851   27315  
1221                Sedan           34        24        2202   25030  
966           Convertible           28        20        1715  200400  
5671        4dr Hatchback           31        21         873   27295  
1497                Coupe           26        18        1624   47750  
5268                Sedan           25        15         238   69800  
9727                Sedan           32        23        1439   28575  
6226        4dr Hatchback           34        28        2009   24830

<hr>

# 3. Checking Data

<a id = "Check_Data"></a>

In [101]:
car_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

> It appears that we have 16 features and 11914 entries. However the dataset seems to be incomplete with 

> We should drop some of the columns as they would not provide sufficient data for us

> The code below shows us the columns with empty data, and the total count of empty data.

In [102]:
null_col = car_raw.columns[car_raw.isnull().any()]

car_to_clean = car_raw

car_raw[null_col].isnull().sum()

Engine Fuel Type       3
Engine HP             69
Engine Cylinders      30
Number of Doors        6
Market Category     3742
dtype: int64

>Checking Engine Fuel Type. Due to imcomplete information we will drop these columns

In [103]:
null = car_to_clean[car_to_clean.isna().any(axis=1)]
car_to_clean[car_to_clean['Engine Fuel Type'].isna()]

Make   Model  Year Engine Fuel Type  Engine HP  Engine Cylinders  \
11321  Suzuki  Verona  2004              NaN      155.0               6.0   
11322  Suzuki  Verona  2004              NaN      155.0               6.0   
11323  Suzuki  Verona  2004              NaN      155.0               6.0   

      Transmission Type      Driven_Wheels  Number of Doors Market Category  \
11321         AUTOMATIC  front wheel drive              4.0             NaN   
11322         AUTOMATIC  front wheel drive              4.0             NaN   
11323         AUTOMATIC  front wheel drive              4.0             NaN   

      Vehicle Size Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
11321      Midsize         Sedan           25        17         481  17199  
11322      Midsize         Sedan           25        17         481  20199  
11323      Midsize         Sedan           25        17         481  18499

In [104]:
car_to_clean.iloc[11324:11327]

Make   Model  Year  Engine Fuel Type  Engine HP  Engine Cylinders  \
11324  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11325  Suzuki  Verona  2005  regular unleaded      155.0               6.0   
11326  Suzuki  Verona  2005  regular unleaded      155.0               6.0   

      Transmission Type      Driven_Wheels  Number of Doors Market Category  \
11324         AUTOMATIC  front wheel drive              4.0             NaN   
11325         AUTOMATIC  front wheel drive              4.0             NaN   
11326         AUTOMATIC  front wheel drive              4.0             NaN   

      Vehicle Size Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
11324      Midsize         Sedan           25        18         481  19349  
11325      Midsize         Sedan           25        18         481  21049  
11326      Midsize         Sedan           25        18         481  17549

Looking at the next 3 cars. They are of similar make and model, just different years. Therefore we will let the 2004 model be regular unleaded.

In [105]:
car_to_clean_map = {label:idx for idx,label in enumerate(car_to_clean["Engine Fuel Type"].unique())}
car_to_clean["Engine Fuel Type"] = car_to_clean["Engine Fuel Type"].map(car_to_clean_map)

car_to_clean.at[11321, 'Engine Fuel Type'] = 1
car_to_clean.at[11322, 'Engine Fuel Type'] = 1
car_to_clean.at[11323, 'Engine Fuel Type'] = 1

car_to_clean.iloc[11321:11324]

Make   Model  Year  Engine Fuel Type  Engine HP  Engine Cylinders  \
11321  Suzuki  Verona  2004                 1      155.0               6.0   
11322  Suzuki  Verona  2004                 1      155.0               6.0   
11323  Suzuki  Verona  2004                 1      155.0               6.0   

      Transmission Type      Driven_Wheels  Number of Doors Market Category  \
11321         AUTOMATIC  front wheel drive              4.0             NaN   
11322         AUTOMATIC  front wheel drive              4.0             NaN   
11323         AUTOMATIC  front wheel drive              4.0             NaN   

      Vehicle Size Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
11321      Midsize         Sedan           25        17         481  17199  
11322      Midsize         Sedan           25        17         481  20199  
11323      Midsize         Sedan           25        17         481  18499

>Checking Engine HP we can see that most electric cars do not have their HP allocated. Therefore we will use the mean to input the data

In [106]:
car_to_clean[car_to_clean['Engine HP'].isna()]

Make        Model  Year  Engine Fuel Type  Engine HP  \
539      FIAT         500e  2015                 5        NaN   
540      FIAT         500e  2016                 5        NaN   
541      FIAT         500e  2017                 5        NaN   
2905  Lincoln  Continental  2017                 2        NaN   
2906  Lincoln  Continental  2017                 2        NaN   
...       ...          ...   ...               ...        ...   
9850      Kia      Soul EV  2015                 5        NaN   
9851      Kia      Soul EV  2015                 5        NaN   
9852      Kia      Soul EV  2016                 5        NaN   
9853      Kia      Soul EV  2016                 5        NaN   
9854      Kia      Soul EV  2016                 5        NaN   

      Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
539                0.0      DIRECT_DRIVE  front wheel drive              2.0   
540                0.0      DIRECT_DRIVE  front wheel drive              2.0   
541                0.0      DIRECT_DRIVE  front wheel drive              2.0   
2905               6.0         AUTOMATIC    all wheel drive              4.0   
2906               6.0         AUTOMATIC  front wheel drive              4.0   
...                ...               ...                ...              ...   
9850               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9851               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9852               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9853               0.0      DIRECT_DRIVE  front wheel drive              4.0   
9854               0.0      DIRECT_DRIVE  front wheel drive              4.0   

     Market Category Vehicle Size  Vehicle Style  highway MPG  city mpg  \
539        Hatchback      Compact  2dr Hatchback          108       122   
540        Hatchback      Compact  2dr Hatchback          103       121   
541        Hatchback      Compact  2dr Hatchback          103       121   
2905          Luxury        Large          Sedan           25        17   
2906          Luxury        Large          Sedan           27        18   
...              ...          ...            ...          ...       ...   
9850             NaN      Compact          Wagon           92       120   
9851             NaN      Compact          Wagon           92       120   
9852             NaN      Compact          Wagon           92       120   
9853             NaN      Compact          Wagon           92       120   
9854             NaN      Compact          Wagon           92       120   

      Popularity   MSRP  
539          819  31800  
540          819  31800  
541          819  31800  
2905          61  55915  
2906          61  62915  
...          ...    ...  
9850        1720  35700  
9851        1720  33700  
9852        1720  33950  
9853        1720  31950  
9854        1720  35950  

[69 rows x 16 columns]

In [107]:
car_to_clean_HP = np.where(car_to_clean['Engine HP'].isnull())

for i in range(len(car_to_clean_HP[0])):
    car_to_clean.loc[car_to_clean_HP[0][i], 'Engine HP'] = np.mean(car_to_clean['Engine HP'])
    
car_to_clean.iloc[540:542]

Make Model  Year  Engine Fuel Type  Engine HP  Engine Cylinders  \
540  FIAT  500e  2016                 5  249.38607               0.0   
541  FIAT  500e  2017                 5  249.38607               0.0   

    Transmission Type      Driven_Wheels  Number of Doors Market Category  \
540      DIRECT_DRIVE  front wheel drive              2.0       Hatchback   
541      DIRECT_DRIVE  front wheel drive              2.0       Hatchback   

    Vehicle Size  Vehicle Style  highway MPG  city mpg  Popularity   MSRP  
540      Compact  2dr Hatchback          103       121         819  31800  
541      Compact  2dr Hatchback          103       121         819  31800

>Checking Engine Cylinders we can see that all electric cars are missing values. This is due to electric vehicles not having engines.

>Another interesting vehicle model that suffers from this is the RX7 and RX8. This is due to the vehicles having rotary engines that do not have cylinders.

> Therefore we will just let the cars Engine Cylinders = 0

In [108]:
car_to_clean_cylin = np.where(car_to_clean['Engine Cylinders'].isna())
car_to_clean[car_to_clean['Engine Cylinders'].isna()]

Make    Model  Year  Engine Fuel Type  Engine HP  \
1983   Chevrolet  Bolt EV  2017                 5  200.00000   
1984   Chevrolet  Bolt EV  2017                 5  200.00000   
3716  Volkswagen   e-Golf  2015                 5  115.00000   
3717  Volkswagen   e-Golf  2015                 5  115.00000   
3718  Volkswagen   e-Golf  2016                 5  115.00000   
3719  Volkswagen   e-Golf  2016                 5  115.00000   
5778  Mitsubishi   i-MiEV  2014                 5  249.38607   
5779  Mitsubishi   i-MiEV  2016                 5   66.00000   
5780  Mitsubishi   i-MiEV  2017                 5   66.00000   
8373      Toyota  RAV4 EV  2012                 5  154.00000   
8695       Mazda     RX-7  1993                 1  255.00000   
8696       Mazda     RX-7  1994                 1  255.00000   
8697       Mazda     RX-7  1995                 1  255.00000   
8698       Mazda     RX-8  2009                 0  232.00000   
8699       Mazda     RX-8  2009                 0  212.00000   
8700       Mazda     RX-8  2009                 0  232.00000   
8701       Mazda     RX-8  2009                 0  232.00000   
8702       Mazda     RX-8  2009                 0  232.00000   
8703       Mazda     RX-8  2009                 0  212.00000   
8704       Mazda     RX-8  2009                 0  212.00000   
8705       Mazda     RX-8  2010                 0  232.00000   
8706       Mazda     RX-8  2010                 0  212.00000   
8707       Mazda     RX-8  2010                 0  212.00000   
8708       Mazda     RX-8  2010                 0  232.00000   
8709       Mazda     RX-8  2010                 0  232.00000   
8710       Mazda     RX-8  2011                 0  212.00000   
8711       Mazda     RX-8  2011                 0  232.00000   
8712       Mazda     RX-8  2011                 0  232.00000   
8713       Mazda     RX-8  2011                 0  232.00000   
8714       Mazda     RX-8  2011                 0  212.00000   

      Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
1983               NaN      DIRECT_DRIVE  front wheel drive              4.0   
1984               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3716               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3717               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3718               NaN      DIRECT_DRIVE  front wheel drive              4.0   
3719               NaN      DIRECT_DRIVE  front wheel drive              4.0   
5778               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5779               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
5780               NaN      DIRECT_DRIVE   rear wheel drive              4.0   
8373               NaN      DIRECT_DRIVE  front wheel drive              4.0   
8695               NaN            MANUAL   rear wheel drive              2.0   
8696               NaN            MANUAL   rear wheel drive              2.0   
8697               NaN            MANUAL   rear wheel drive              2.0   
8698               NaN            MANUAL   rear wheel drive              4.0   
8699               NaN         AUTOMATIC   rear wheel drive              4.0   
8700               NaN            MANUAL   rear wheel drive              4.0   
8701               NaN            MANUAL   rear wheel drive              4.0   
8702               NaN            MANUAL   rear wheel drive              4.0   
8703               NaN         AUTOMATIC   rear wheel drive              4.0   
8704               NaN         AUTOMATIC   rear wheel drive              4.0   
8705               NaN            MANUAL   rear wheel drive              4.0   
8706               NaN         AUTOMATIC   rear wheel drive              4.0   
8707               NaN         AUTOMATIC   rear wheel drive              4.0   
8708               NaN            MANUAL   rear wheel drive              4.0   
8709               NaN     

In [109]:
car_to_clean.loc[car_to_clean_cylin[0][0:30], 'Engine Cylinders'] = 0
car_to_clean[car_to_clean['Engine Cylinders'].isna()]

Empty DataFrame
Columns: [Make, Model, Year, Engine Fuel Type, Engine HP, Engine Cylinders, Transmission Type, Driven_Wheels, Number of Doors, Market Category, Vehicle Size, Vehicle Style, highway MPG, city mpg, Popularity, MSRP]
Index: []

>Checking the missing values. Due to the small sample size of this column we will manually check for the information online and input them into the dataframe

>[Ferrari FF](https://en.wikipedia.org/wiki/Ferrari_FF) __Has 3 doors__

>[Tesla Model S](https://en.wikipedia.org/wiki/Tesla_Model_S) __Has 4 doors__

In [110]:
car_to_clean[car_to_clean['Number of Doors'].isna()]

Make    Model  Year  Engine Fuel Type  Engine HP  Engine Cylinders  \
4666  Ferrari       FF  2013                 0  651.00000              12.0   
6930    Tesla  Model S  2016                 5  249.38607               0.0   
6931    Tesla  Model S  2016                 5  249.38607               0.0   
6932    Tesla  Model S  2016                 5  249.38607               0.0   
6933    Tesla  Model S  2016                 5  249.38607               0.0   
6934    Tesla  Model S  2016                 5  249.38607               0.0   

     Transmission Type     Driven_Wheels  Number of Doors  \
4666  AUTOMATED_MANUAL   all wheel drive              NaN   
6930      DIRECT_DRIVE   all wheel drive              NaN   
6931      DIRECT_DRIVE   all wheel drive              NaN   
6932      DIRECT_DRIVE   all wheel drive              NaN   
6933      DIRECT_DRIVE  rear wheel drive              NaN   
6934      DIRECT_DRIVE   all wheel drive              NaN   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
4666  Exotic,High-Performance        Large         Coupe           16   
6930       Exotic,Performance        Large         Sedan          105   
6931       Exotic,Performance        Large         Sedan          101   
6932  Exotic,High-Performance        Large         Sedan          105   
6933       Exotic,Performance        Large         Sedan          100   
6934       Exotic,Performance        Large         Sedan          107   

      city mpg  Popularity    MSRP  
4666        11        2774  295000  
6930       102        1391   79500  
6931        98        1391   66000  
6932        92        1391  134500  
6933        97        1391   74500  
6934       101        1391   71000

In [111]:
car_to_clean.loc[4666, 'Number of Doors'] = 3
car_to_clean.loc[6930, 'Number of Doors'] = 4
car_to_clean.loc[6931, 'Number of Doors'] = 4
car_to_clean.loc[6932, 'Number of Doors'] = 4
car_to_clean.loc[6933, 'Number of Doors'] = 4
car_to_clean.loc[6934, 'Number of Doors'] = 4

> Due to the sheer size of the missing values in the "Market Category" columns we will categorize the missing values as No Cat

In [112]:
car_to_clean_mark = np.where(car_to_clean['Market Category'].isna())
car_to_clean.loc[car_to_clean_mark[0], 'Market Category'] = 'No Cat' 

<hr>

# 4. Performing EDA

<a id = "EDA"></a>